### PR TITLE
feat: optimize AsyncFetch cache with init size

### PR DESCRIPTION
**Pull request by fudongyingluck**
At most of times, we know the cache size of the AsyncFetch. But we don't use this information for speed up the cache allocation. When we have lots of nodes, cache resize is time consuming. For example, when we have 200 nodes, and with 0.75 HashMap load factor, we might need more than 7 times HashMap resize to get an array with length 512.
In this change, we will init cache with size of nodes that we know already. It's might not very correct but can't be worse, I think

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java [hunk 1]
```diff
@@ -53,19 +53,20 @@
...
 
     @SuppressWarnings("unchecked")
-    protected AsyncShardFetch(Logger logger, String type, ShardId shardId, String customDataPath) {
+    protected AsyncShardFetch(Logger logger, String type, ShardId shardId, String customDataPath, int nodeNumber) {
         this.logger = logger;
         this.type = type;
         this.shardId = Objects.requireNonNull(shardId);
         this.customDataPath = Objects.requireNonNull(customDataPath);
         this.fetchingCount = 0;
+        this.cache = new HashMap<>(nodeNumber);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:69]: This actually isn't quite large enough, you should use `org.elasticsearch.common.util.Maps#newMapWithExpectedSize` here.

File: server/src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java [hunk 2]
```diff
@@ -53,19 +53,20 @@
...
-    private final Map<String, NodeEntry<T>> cache = new HashMap<>();
+    private final Map<String, NodeEntry<T>> cache;
     private final Set<String> nodesToIgnore = new HashSet<>();
     private final AtomicLong round = new AtomicLong();
     private boolean closed;
     private volatile int fetchingCount;
 
     @SuppressWarnings("unchecked")
-    protected AsyncShardFetch(Logger logger, String type, ShardId shardId, String customDataPath) {
+    protected AsyncShardFetch(Logger logger, String type, ShardId shardId, String customDataPath, int nodeNumber) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:63]: naming nit/suggestion:

```suggestion
    protected AsyncShardFetch(Logger logger, String type, ShardId shardId, String customDataPath, int expectedSize) {
```

File: server/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java [hunk 3]
```diff
@@ -159,7 +159,10 @@ private class AsyncShardStoresInfoFetches {
         void start() {
             try {
                 for (Tuple<ShardId, String> shard : shards) {
-                    new InternalAsyncFetch(logger, "shard_stores", shard.v1(), shard.v2()).fetchData(nodes, Collections.emptySet());
+                    new InternalAsyncFetch(logger, "shard_stores", shard.v1(), shard.v2(), this.nodes.getSize()).fetchData(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java:162]: We only need allocate enough to hold all the data nodes so `routingNodes.size()` is better here:

```suggestion
                    new InternalAsyncFetch(logger, "shard_stores", shard.v1(), shard.v2(), routingNodes.size()).fetchData(
```

File: server/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java [hunk 4]
```diff
@@ -184,8 +187,8 @@ private class InternalAsyncFetch extends AsyncShardFetch<NodeGatewayStartedShard
 
             private final Releasable ref = refs.acquire();
 
-            InternalAsyncFetch(Logger logger, String type, ShardId shardId, String customDataPath) {
-                super(logger, type, shardId, customDataPath);
+            InternalAsyncFetch(Logger logger, String type, ShardId shardId, String customDataPath, int nodeNumber) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java:190]: naming nit/suggestion

```suggestion
            InternalAsyncFetch(Logger logger, String type, ShardId shardId, String customDataPath, int expectedSize) {
```

File: server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 5]
```diff
@@ -213,8 +213,8 @@ private boolean hasNewNodes(DiscoveryNodes nodes) {
 
     abstract class InternalAsyncFetch<T extends BaseNodeResponse> extends AsyncShardFetch<T> {
 
-        InternalAsyncFetch(Logger logger, String type, ShardId shardId, String customDataPath) {
-            super(logger, type, shardId, customDataPath);
+        InternalAsyncFetch(Logger logger, String type, ShardId shardId, String customDataPath, int nodeNumber) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:216]: naming nit/suggestion

```suggestion
        InternalAsyncFetch(Logger logger, String type, ShardId shardId, String customDataPath, int expectedSize) {
```

File: server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 6]
```diff
@@ -249,7 +249,8 @@ protected AsyncShardFetch.FetchResult<NodeGatewayStartedShards> fetchData(ShardR
                     logger,
                     "shard_started",
                     shardId,
-                    IndexMetadata.INDEX_DATA_PATH_SETTING.get(allocation.metadata().index(shard.index()).getSettings())
+                    IndexMetadata.INDEX_DATA_PATH_SETTING.get(allocation.metadata().index(shard.index()).getSettings()),
+                    allocation.nodes().getSize()
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:253]: Just the data nodes:

```suggestion
                    allocation.routingNodes().size()
```

File: server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 7]
```diff
@@ -295,7 +296,8 @@ protected AsyncShardFetch.FetchResult<NodeStoreFilesMetadata> fetchData(ShardRou
                     logger,
                     "shard_store",
                     shard.shardId(),
-                    IndexMetadata.INDEX_DATA_PATH_SETTING.get(allocation.metadata().index(shard.index()).getSettings())
+                    IndexMetadata.INDEX_DATA_PATH_SETTING.get(allocation.metadata().index(shard.index()).getSettings()),
+                    allocation.nodes().getSize()
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:300]: Just the data nodes:

```suggestion
                    allocation.routingNodes().size()
```

**Review by DaveCTurner — COMMENTED** [review]
Makes sense to me; I left a few comments.

**Comment by fudongyingluck** [comment]
Thanks @DaveCTurner for your time and patient to a newbie in java, es, and github ^_^

**Comment by DaveCTurner** [comment]
Looks good; please don't force-push to PR branches tho, it makes it harder to keep track of reviews and other context.

@elasticmachine ok to test

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
Thanks for the contribution, @fudongyingluck. Great work.

**Pull request merged**